### PR TITLE
DocumentationComment.py: Accept `position`

### PR DESCRIPTION
**Pull request by damngamerz**
DocumentationComment accept `position` instead
of `range` in `__init__`.
Amend tests to comply with changes.
Closes https://github.com/coala/coala/issues/2646

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 1]
```diff
@@ -32,14 +33,20 @@ def __init__(self, documentation, docstyle_definition,
...
+            The starting ``TextPosition`` of the documentation.
         """
         self.documentation = documentation
         self.docstyle_definition = docstyle_definition
         self.indent = indent
         self.marker = marker
-        self.range = range
+        self.position = position
+        self.assemble = None if position is None else self._assemble()
+        self.range = None if position is None else TextRange.from_values(
```
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: PS- This is important. As if directly implemented will cause an `AttributeError: 'NoneType' object has no attribute`

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def test_fields(self):
         self.assertEqual(uut.indent, ' ')
         self.assertEqual(str(uut), 'my doc')
         self.assertEqual(uut.marker, ('/**', '*', '*/'))
-        self.assertEqual(uut.range, (25, 45))
+        self.assertEqual(uut.position, (25, 45))
```
Comment (damngamerz) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: This is test fails. I can't understand the objective of `def test_fields()`. Hence, I was unable to resolve it.
CC @Makman2 @NiklasMM

Comment (Makman2) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: Imo it shall just test if fields/properties are populated correctly from `__init__` calls.

**Comment by damngamerz** [comment]
@Makman2 ready for review!!!

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 3]
```diff
@@ -32,14 +33,20 @@ def __init__(self, documentation, docstyle_definition,
...
+        :param position:
+            The starting ``TextPosition`` of the documentation.
         """
         self.documentation = documentation
         self.docstyle_definition = docstyle_definition
         self.indent = indent
         self.marker = marker
-        self.range = range
+        self.position = position
+        self.assemble = None if position is None else self._assemble()
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: 1. Call it `assembled` at least as it's a field, not a function :)
2. Do we need this? Imo you can also assemble a docstring without a `range` or `position`

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 1 above)
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: That's fine :)

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 3 above)
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: If you like this value to be cached, implement a property `assembled` with `@lru_cache`. I hope this value can't get outdated^^

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py [hunk 4]
```diff
@@ -56,7 +57,7 @@ def test_fields(self):
         self.assertEqual(uut.indent, '\t')
         self.assertEqual(str(uut), 'qwertzuiop')
         self.assertEqual(uut.marker, ('##', '#', '#'))
-        self.assertEqual(uut.range, None)
```
Comment (Makman2) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: Isn't this one `None` too? would keep the assert then, doesn't harm :)

Comment (damngamerz) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]:
```
uut = DocumentationComment('qwertzuiop',
                                    python_doxygen,
                                    '\t',
                                    ('##', '#', '#'),
                                    None)
```
Yup it is but assertion will fail cause now we accept `position` :tada:

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 3 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: oh yes I tried it implementing it as a field bounding `self.assemble` to `_assemble()` but the problem is we need that final `assembled` docstring which `_assembled()` returns, as for constructing the `range`. Cause functions like `.count()` ,`.rfind()` works only on strings.

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Btw i liked the idea of `@lru_cache` Im implementing it. It will be good when later part (bear) calls it and it will never get outdated YAY! :wink`

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: In `@lru_cache` we have to compulsorily provide `maxsize` argument. Our docstrings are dynamic in nature hence, setting it to `None`. And this saves memory yes!!

Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Don't let the cache grow into nowhere with `maxsize=None`, use a sensible size like `64` or `128`. Though as `assemble()` takes no parameters, `lru_cache` will only cache a single item, thus we can set `maxsize=1` and use `assemble.clear_cache` when modifications occur^^
(Though this is error prone, as we might forget to reset the cache properly^^)

File: tests/bearlib/languages/documentation/DocumentationCommentTest.py (hunk 4 above)
Comment (Makman2) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: but the field is still available on the instantiated object, isn't it?

Comment (damngamerz) [inline tests/bearlib/languages/documentation/DocumentationCommentTest.py]: Yup^ ok let me revert it back and add a separate assert field for `position`

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 3 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: amending changes :+1:

**Comment by damngamerz** [comment]
@Makman2 ready to be reviewed again :+1:

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 3 above)
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: 1. As you have an `lru_cache` you can keep `self.assemble()` like it is without renaming it to `self._assemble()`.
2. Also you don't need the position to assemble a docstring.
3. What if I change e.g. markers now? Then we need to update `self.assemble()` again. Or do we consider those fields to be "immutable"?

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: >As you have an lru_cache you can keep self.assemble() like it is without renaming it to self._assemble().

Yup we can now let me amend it and see.
>What if I change e.g. markers now? Then we need to update self.assemble() again. Or do we consider those fields to be "immutable"?

Hmm I don't know. My possible guess is yes we have to update`self.assemble()` again. I don't see any reason for being immutable.

Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: This is what I meant with
> (Though this is error prone, as we might forget to reset the cache properly^^)

:D

Up to you how you want to treat the fields. They are currently not prefixed with `_` identifying them as "private" (or "protected" or whatsoever), so from outside they look like being settable.

Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: yup^ although there's a whole revamp coming up in in PHASE-2 when `docbaseclass()` will be implemented. This can be taken care in that^ :wink: .

what i meant by that is the `assemble()` part of `docbaseclass()` we can reset the cache properly there so as to make sure wherever it is called a fresh assembled docstring is fetched. for internal purpose we can use the cached `assemble()` :D

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 5]
```diff
@@ -32,14 +34,20 @@ def __init__(self, documentation, docstyle_definition,
...
+        :param position:
+            The starting ``TextPosition`` of the documentation.
         """
         self.documentation = documentation
         self.docstyle_definition = docstyle_definition
         self.indent = indent
         self.marker = marker
-        self.range = range
+        self.position = position
+        self.assemble = None if position is None else self.assemble()
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: I still don't like this, why do we need this?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 6]
```diff
@@ -193,16 +201,16 @@ def from_metadata(cls, doccomment, docstyle_definition,
...
         >>> Description = DocumentationComment.Description
         >>> Parameter = DocumentationComment.Parameter
         >>> python_default = DocstyleDefinition.load("python3", "default")
         >>> parsed_doc = [Description(desc='\nDescription\n'),
         ...               Parameter(name='age', desc=' Age\n')]
         >>> str(DocumentationComment.from_metadata(
         ...         parsed_doc, python_default,
-        ...         python_default.markers[0], 4,
-        ...         TextRange.from_values(0, 0, 0, 0)))
+        ...         python_default.markers[0], "    ",
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: 1. Use single quotes if possible :)
2. Shouldn't this go into an own commit?

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 7]
```diff
@@ -239,8 +247,9 @@ def from_metadata(cls, doccomment, docstyle_definition,
             assembled_doc += ''.join(section_desc)
 
         return DocumentationComment(assembled_doc, docstyle_definition, indent,
-                                    marker, range)
+                                    marker, position)
 
+    @lru_cache(maxsize=1)
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: we don't reset the cache now, so once `assemble` is executed, the result won't change any more. Are you planning to reset the cache when properties change?

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 5 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: I guess you're asking for `self.position` yep it's required. Is currently being used in tests and later in `DocStyleBear`. 
For `self.assemble` it's being used to construct that later `self.range`

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 6 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Amended. Separate commit. hmm this caused that `pydocbear` test failure. separate commit it is!

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 7 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: Yep `docbaseclass()` as I explained here https://github.com/coala/coala/pull/4368#discussion_r123424460

File: coalib/bearlib/languages/documentation/DocumentationComment.py [hunk 8]
```diff
@@ -32,14 +34,20 @@ def __init__(self, documentation, docstyle_definition,
...
+        :param position:
+            The starting ``TextPosition`` of the documentation.
         """
         self.documentation = documentation
         self.docstyle_definition = docstyle_definition
         self.indent = indent
         self.marker = marker
-        self.range = range
+        self.position = position
+        self.assemble = self.assemble()
```
Comment (Makman2) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: you are already using `@lru_cache`, don't do this :)

**Comment by Makman2** [comment]
Is the commit order correct? So first `Pass indent`, then `Add conditional None` and then the `Accept position` one?

File: coalib/bearlib/languages/documentation/DocumentationComment.py (hunk 8 above)
Comment (damngamerz) [inline coalib/bearlib/languages/documentation/DocumentationComment.py]: @Makman2 do you want me to remove `self.assemble` ? and use `self.assemble()` instead ?

**Comment by damngamerz** [comment]
@SanketDG This is one ready too :+1:

**Comment by SanketDG** [comment]
ack efd00bb

**Comment by SanketDG** [comment]
ack bac0c3e 526af97

**Comment by SanketDG** [comment]
ack f3f613a

**Comment by SanketDG** [comment]
@rultor merge

**Comment by SanketDG** [comment]
ack f3f613a

**Comment by SanketDG** [comment]
@rultor merge
